### ISSUE TITLE
localkube intermittent segfaults

### ISSUE THREAD

**Issue by gtirloni** [kubernetes/minikube#1847]
**Environment**:

- **Minikube version**: 0.21.0 (compiled from HEAD, 77ea1bad70e379012575e3dcd03e268ce503e3f7)
- **OS**: Fedora 26 (kernel 4.12.5)
- **VM Driver**: VirtualBox 5.1.26
- **ISO version**: 0.23.1

**What happened**:

The cluster fails to start while localkube is segfaulting. Sometimes it settles after 30 seconds, other times it takes 2 minutes to get a ready status from `kubectl get nodes`. Sometimes it works right away.

**What you expected to happen**:

Cluster is available shortly after `minikube start` finishes.

**How to reproduce it**:

The issue is intermittent on my system (roughly 2 out of 3):

```
$ minikube start && while [ true ]; do kubectl get nodes | grep -v NAME; sleep 1; done
Deleting local Kubernetes cluster...
Machine deleted.
Starting local Kubernetes v1.7.3 cluster...
Starting VM...
Getting VM IP address...
Moving files into cluster...
Setting up certs...
Starting cluster components...
Connecting to cluster...
Setting up kubeconfig...
Kubectl is now configured to use the cluster.
No resources found.
No resources found.
No resources found.
Unable to connect to the server: read tcp 192.168.99.1:55014->192.168.99.108:8443: read: connection reset by peer
The connection to the server 192.168.99.108:8443 was refused - did you specify the right host or port?
The connection to the server 192.168.99.108:8443 was refused - did you specify the right host or port?
The connection to the server 192.168.99.108:8443 was refused - did you specify the right host or port?
The connection to the server 192.168.99.108:8443 was refused - did you specify the right host or port?
The connection to the server 192.168.99.108:8443 was refused - did you specify the right host or port?
No resources found.
No resources found.
No resources found.
Unable to connect to the server: read tcp 192.168.99.1:55050->192.168.99.108:8443: read: connection reset by peer
The connection to the server 192.168.99.108:8443 was refused - did you specify the right host or port?
The connection to the server 192.168.99.108:8443 was refused - did you specify the right host or port?
The connection to the server 192.168.99.108:8443 was refused - did you specify the right host or port?
The connection to the server 192.168.99.108:8443 was refused - did you specify the right host or port?
The connection to the server 192.168.99.108:8443 was refused - did you specify the right host or port?
No resources found.
No resources found.
No resources found.
No resources found.
minikube   Ready     1s        v1.7.3
minikube   Ready     2s        v1.7.3
minikube   Ready     3s        v1.7.3
minikube   Ready     4s        v1.7.3
minikube   Ready     5s        v1.7.3
minikube   Ready     6s        v1.7.3
minikube   Ready     7s        v1.7.3
minikube   Ready     8s        v1.7.3
minikube   Ready     9s        v1.7.3
minikube   Ready     10s       v1.7.3
minikube   Ready     12s       v1.7.3
minikube   Ready     13s       v1.7.3
minikube   Ready     14s       v1.7.3
minikube   Ready     15s       v1.7.3
^C
```
Segfault always happens here:

```
Aug 19 19:31:43 minikube localkube[3768]: Starting kubelet...
Aug 19 19:31:43 minikube localkube[3768]: Waiting for kubelet to be healthy...
Aug 19 19:31:43 minikube localkube[3768]: I0819 19:31:43.894767    3768 feature_gate.go:144] feature gates: map[]
Aug 19 19:31:43 minikube localkube[3768]: E0819 19:31:43.894814    3768 server.go:318] unable to register configz: register config "componentconfig" twice
Aug 19 19:31:44 minikube localkube[3768]: I0819 19:31:44.073629    3768 client.go:72] Connecting to docker on unix:///var/run/docker.sock
Aug 19 19:31:44 minikube localkube[3768]: I0819 19:31:44.073652    3768 client.go:92] Start docker client with request timeout=2m0s
Aug 19 19:31:44 minikube localkube[3768]: fatal error: unexpected signal during runtime execution
Aug 19 19:31:44 minikube localkube[3768]: [signal SIGSEGV: segmentation violation code=0x1 addr=0xe5 pc=0x7f84cf71c100]
```

**Output of `minikube logs` (if applicable)**:

[localkube_segfault.txt](https://github.com/kubernetes/minikube/files/1236602/localkube_segfault.txt)

**Anything else do we need to know**:
Cannot reproduce issue with official minukube 0.21.0 (ISO 0.23.0, k8s 1.7.0). Cluster is up after 3-4 failed `kubectl get nodes` and localkube never segfaults.

**Comment by r2d4** [comment]
We have two small patches that we maintain for localkube, this is the one you're encountering. 

https://github.com/kubernetes/minikube/blob/master/hack/kube-proxy-patch.diff

Basically, since localkube runs all the components in a single process, there is some accidental API interaction.  In this case, `configz.New("componentconfig")` is called on all components.  While this used to not be a fatal error, it was recently added as a fatal error in 1.7.  Rather than try to upstream this, we're in the process of moving away from localkube to kubeadm or an alternative bootstrapper.

There is also a TPR patch, which is probably less important as TPRs are being phased out in favor of CRDs, https://github.com/kubernetes/minikube/blob/master/hack/tpr-patch.diff

Neither of these patches really affect the behavior of localkube, but will prevent the race conditioned panics here.

**Comment by r2d4** [comment]
This shouldn't be happened at head, since the vendored dependencies do have these patches.  Are you using the vendored dependencies?  If not, or if you've changed them, be sure to be using `hack/godeps/godep-save.sh` and `hack/godeps/godep-restore.sh` since those should apply these patches as well as a few other vendoring gotchas.

**Comment by gtirloni** [comment]
@r2d4 Thanks for the background information on this issue. I was just running the head version out of habit, not doing any development. I've compiled it from a fresh clone, so no changes from my side. I also confirmed those files were already patched in the vendor directory.

I'm fine running the latest released version while the work on the kubeadm migration is going on.

**Comment by r2d4** [comment]
I'm going to reopen this, because I'm able to reproduce it on a few subsequent starts.

**Comment by r2d4** [comment]
This should be fixed by #1848
